plugins/openai_plugin.py

```diff
@@ -133,143 +178,160 @@ def request_http(self, query: dict, user_id: int, test_end_time: float = 0):
-                        # total tokens before the timeout will be equal to the total tokens in the response.
-                        if not result.output_tokens_before_timeout:
-                            result.output_tokens_before_timeout = result.output_tokens
-
-                except KeyError:
-                    logger.exception("KeyError, unexpected response format in line: %s", line)
+                recv_time = time.time() # Record time asap
+                # Only record lines with data
+                if line:
+                    logger.debug("response line: %s", line)
+                    resps.append(dict(
+                        time = recv_time,
+                        data = line
+                    ))
+            # Full response received
+            result.end_time = time.time()
         except requests.exceptions.ChunkedEncodingError as err:
             result.end_time = time.time()
             result.error_text = repr(err)
-            result.output_text = "".join(tokens)
-            result.output_tokens = len(tokens)
+            #result.output_text = "".join([])
+            result.output_tokens = len(resps)
             if response is not None:
                 result.error_code = response.status_code
             logger.exception("ChunkedEncodingError while streaming response")
             return result
 
+        ###########################################
+        # DO NOT CALL time.time BEYOND THIS POINT #
+        ###########################################
+
+        # If no data was received return early
+        if not resps:
+            result.output_tokens = 0
+            result.error_code = response.status_code
+            return result
+
+        # Check for end of request marker
+        if resps[-1]['data'] == b"data: [DONE]":
+            result.end_time = resps[-1]['time']
+            resps.pop() # Drop the end indicator
+        else:
+            # TODO This signals that the request is incomplete
+            pass
+
+        # Check for usage statistics
+        message = self._process_resp(resps[-1]['data'])
+        # If stream_options.include_usage == True then the final
+        # message contains only token stats
+        expected_output_tokens = None
+        if message and not message.get("choices") and message.get('usage'):
+            # We want to count output tokens ourselves, but we can check our work with usage data.
+            expected_output_tokens = deepget(message, "usage", "completion_tokens")
+            result.input_tokens = deepget(message, "usage", "prompt_tokens")
+            # We don't want to record this message
+            resps.pop()
+
+        # Iterate through all responses
+        tokens = []
+        prev_time = 0
+        first_token_seen = False
+        for resp in resps:
+            message = self._process_resp(resp['data'])
+            if not message:
+                # TODO: This may be bad
+                continue
```

Comment:
If a token has an error in processing, we log it and continue. Likewise to the previous comment do we want an error state for this?

Reply:
I guess it is okay to skip over the messages due to failed requests as long as we log the failure status and reason in our log file for that specific request.